File: app/src/main/java/com/iamwent/diary/utils/LunarUtil.java
package com.iamwent.diary.utils;

import java.util.Calendar;

/**
 * Created by iamwent on 9/19/16.
 *
 * @author iamwent
 * @since 9/19/16
 */
public class LunarUtil {

    private static final String[] numbers = {"零", "一", "二", "三", "四", "五", "六", "七", "八", "九", "十"};

    public static String time2Chinese(long timeInMillis) {
        Calendar c = Calendar.getInstance();
        c.setTimeInMillis(timeInMillis);

        String year = year2Chinese(c.get(Calendar.YEAR));
        String month = month2Chinese(c.get(Calendar.MONTH));
        String day = day2Chinese(c.get(Calendar.DATE));

        return String.format("%s年 %s月 %s日", year, month, day);
    }

    public static String year2Chinese(int year) {
        String chinese = "";

        do {
            int tail = year % 10;
            year /= 10;

            chinese = numbers[tail] + chinese;
        } while (year > 0);

        return chinese;
    }

    public static String month2Chinese(int month) {
        month = month + 1;

        String chinese;
        if (month > 10) {
            chinese = numbers[10] + numbers[month % 10];
        } else if (month == 10) {
            chinese = numbers[10];
        } else {
            chinese = numbers[month];
        }

        return chinese;
    }

    public static String day2Chinese(int day) {
        String chinese;
        if (day % 10 == 0) {
            if (day == 10) {
                chinese = numbers[day];
            } else {
                chinese = numbers[day / 10] + numbers[10];
            }
        } else {
            if (day > 20) {
                chinese = numbers[day / 10] + numbers[10] + numbers[day % 10];
            } else if (day > 10) {
                chinese = numbers[10] + numbers[day % 10];
            } else {
                chinese = numbers[day];
            }
        }

        return chinese;
    }
}

File: app/src/main/java/com/iamwent/diary/utils/LunarUtil.kt
package com.iamwent.diary.utils

import java.time.LocalDateTime

/**
 * Created by iamwent on 9/19/16.
 *
 * @author iamwent
 * @since 9/19/16
 */
object LunarUtil {
    private val numbers = arrayOf(
        "零", "一", "二", "三", "四", "五", "六", "七", "八", "九",
        "十", "十一", "十二", "十三", "十四", "十五", "十六", "十七", "十八", "十九",
        "廿", "廿一", "廿二", "廿三", "廿四", "廿五", "廿六", "廿七", "廿八", "廿九",
        "卅", "卅一",
    )

    fun time2Chinese(dateTime: LocalDateTime): String {
        val year = year2Chinese(dateTime.year)
        val month = month2Chinese(dateTime.monthValue)
        val day = day2Chinese(dateTime.dayOfMonth)
        return "${year}年 ${month}月 ${day}日"
    }

    fun year2Chinese(year: Int): String {
        check(year in 1900..9999) {
            "illegal year range [1900-9999]: $year"
        }

        val builder = StringBuilder()
        var tail = year
        do {
            builder.append(numbers[tail % 10])
            tail /= 10
        } while (tail > 0)
        return builder.reverse().toString()
    }

    fun month2Chinese(month: Int): String {
        check(month in (1..12)) {
            "illegal month range [1-12]: $month"
        }
        return numbers[month]
    }

    fun day2Chinese(day: Int): String {
        check(day in 1..31) {
            "illegal day range [1-31]: $day"
        }
        return numbers[day]
    }
}
